Add toppra

Skwarson96 opened the pull request:
Hello as in the  https://github.com/sea-bass/pyroboplan/issues/55 I added toppra to the cartesian_path.py script I hope it was something like this. I had a problem with the numpy version, because it worked for version 1.23.0

sea-bass commented on examples/cartesian_path.py:
```diff
@@ -8,6 +8,9 @@
 import pinocchio
 from pinocchio.visualize import MeshcatVisualizer
 from scipy.spatial.transform import Rotation
+import toppra as ta
+import toppra.constraint as constraint
+import toppra.algorithm as algo
```
would prefer not to alias these imports if possible

sea-bass commented on examples/cartesian_path.py:
```diff
@@ -76,16 +81,36 @@
+alims = np.ones(q_start.shape[0]) * max_linear_acceleration
+# Define the geometric path and two constraints.
+path = ta.SplineInterpolator(t_vec, q_vec.T)
+pc_vel = constraint.JointVelocityConstraint(vlims)
+pc_acc = constraint.JointAccelerationConstraint(alims)
```
I don't think you want to be using Cartesian velocity limits here, right?

You probably want to be using the *joint* velocity limits from the Pinocchio model. And acceleration is something we have to make up since it's not part of the URDF.

For example, I changed the code to say

```python
vlims = model.velocityLimit
alims = 10* np.ones_like(vlims)
```

and now the trajectories are actually *faster*, not slower, with TOPP-RA.

![image](https://github.com/user-attachments/assets/9b9706ac-f5ad-4937-989a-49bec8469e61)

Skwarson96 replied:
Yeah, thanks. I will check that

sea-bass reviewed (changes requested):
Thanks for doing this! However, I found a big issue.

TOPP-RA will fail to import properly with an error similar to this https://github.com/hungpham2511/toppra/issues/206 -- I isolated it to using NumPy version 2.x, and I could only resolve this problem by downgrading my NumPy with `pip install "numpy<2.0"` ... I don't know how I feel about downgrading NumPy that far back, so at least this may require some strong documentation about downgrading only if you want TOPP-RA.

Related to that, the use of TOPP-RA for this example should be optional, so we should also add some kind of boolean flag to control this.

sea-bass commented on examples/cartesian_path.py:
```diff
@@ -76,16 +81,36 @@
+instance = algo.TOPPRA([pc_vel, pc_acc], path, parametrizer="ParametrizeConstAccel")
+jnt_traj = instance.compute_trajectory()
+
+ts_sample = np.linspace(0, jnt_traj.duration, 100)
+qs_sample = jnt_traj(ts_sample)
```
One more thing -- the animation after you press "Enter" still uses the joint trajectory before time parameterization with TOPP-RA. You probably want to change it to use this new retimed version!

Skwarson96 replied:
Ok, I will change that.

sea-bass commented:
... actually, even better. I just found @zkingston 's fork of TOPPRA at https://github.com/CoMMALab/toppra, and by pulling down this repo and installing it with `pip install -e .`, your example works.

Thanks, Zak!

zkingston commented:
Glad that this was helpful, lol. I opened a PR on TOPPRA (https://github.com/hungpham2511/toppra/pull/259) as well.

Skwarson96 commented:
I will try to address the numpy library issue tomorrow

sea-bass commented:
> Glad that this was helpful, lol. I opened a PR on TOPPRA ([hungpham2511/toppra#259](https://github.com/hungpham2511/toppra/pull/259)) as well.

Thanks! TOPP-RA repo hasn't merged any PRs in 2 years, so I think using your fork is probably a safer bet... but you never know.

Skwarson96 commented:
What I updated:
*  added boolean flag to deciding whether to use TOPP-RA. 
* Added TOPP-RA fork https://github.com/CoMMALab/toppra/tree/develop as submodule
* Update readme.md

sea-bass commented on examples/cartesian_path.py:
```diff
@@ -67,7 +67,10 @@
     max_angular_acceleration=1.0,
 )
 
+TOPP_RA = False
```
nonstandard variable name, maybe call it `use_toppra`

sea-bass commented on README.md:
````diff
@@ -45,3 +45,11 @@ Install this package and its dependencies.
 pip3 install -e .
 ```
+#### TOPP-RA (optional)
+If you want to use the [TOPP-RA](https://github.com/CoMMALab/toppra/tree/develop) repository for path optimization (implemented in examples/cartesian_path.py).
````
```suggestion
If you want to use the [TOPP-RA](https://github.com/CoMMALab/toppra/tree/develop) repository for path optimization (implemented in `examples/cartesian_path.py`).

sea-bass commented on examples/cartesian_path.py:
```diff
@@ -78,13 +81,34 @@
+
+    # replan path with TOPP-RA
+    vlims = model.velocityLimit
+    alims = 10 * np.ones_like(vlims)
```
I think this was just an example I used temporarily, but we probably need to add comments here.

Namely that URDF does not support acceleration limits, so you have to define this information separately and we're just using some provisional values.

At some point, I intend to add a "wrapper" around the Pinocchio model where acceleration/jerk limits can be specified, so a temporary comment like the above is OK for now.

Skwarson96 replied:
Will such a comment be ok?
```suggestion
    # URDF does not support acceleration limits, you must specify this value yourself or use this temporary one
    alims = 10 * np.ones_like(vlims)
```

sea-bass replied:
That works for now, thanks!

sea-bass reviewed (commented):
Just a few small comments.

I don't particularly like that the TOPP-RA dependency has to come in as a submodule, but you did a great job at making it optional (including the README and importing only if enabled). So this will do just fine until/unless the NumPy fix PR is merged and the PyPi version of TOPP-RA is released.

Thank you -- almost there!
